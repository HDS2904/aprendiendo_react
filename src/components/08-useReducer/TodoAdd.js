import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ( { handleAdd } ) => {
	const [ { description }, handleInputChange, reset ] = useForm({ description: ''});

	const handleSubmit = (e) => {
		e.preventDefault();

		if(e.description?.trim().length <= 0){
			return;
		}

		const newTodo = {
			id: new Date().getTime(),
			desc: description,
			done: false
		}

		const action = {
			type: 'add',
			payload: newTodo
		}

		handleAdd(action)
		reset();
	}

	return (
		<>
			<h4>Agregar TODO</h4>
			<hr />
				<form onSubmit={handleSubmit}>
					<input 
						className='form-control'
						type="text" 
						name="description"
						autoComplete='off'
						placeholder='Actividad...'
						value={description}
						onChange={handleInputChange}
					/>
					<button
						className='btn btn-outline-primary mt-3 btn-block'
						type='submit'
					>
						Agregar
					</button>
				</form>
		</>
	)
}
