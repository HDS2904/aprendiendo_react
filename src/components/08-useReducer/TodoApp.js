import React, { useReducer } from 'react'
import { todoReducer } from './todoReducer'
import './style.css'

const initialState = [{
	id: new Date().getTime(),
	desc: 'Aprendiendo React',
	done: false
}]

export const TodoApp = () => {
	const [ todos, dispatch ] = useReducer(todoReducer, initialState)
	console.log(todos);
	
	const handleSubmit = (e) => {
		e.preventDefault();

		const newTodo = {
			id: new Date().getTime(),
			desc: 'Nueva actividad',
			done: false
		}

		const action = {
			type: 'add',
			payload: newTodo
		}

		dispatch(action);
	}

	return (
		<div>
			<h1>TodoApp ({todos.length}) </h1>
			<hr />
			<div className="row">
				<div className="col-7">
					<ul className='list-group list-group-flush'> 
						{
							todos.map((todo, index) => (
								<li
									className='list-group-item'
									key={todo.id}
								>
									<p> { index + 1 }. { todo.desc } </p>
									<button
										className='btn btn-danger'
									>
										Borrar
									</button>
								</li>
							))
						}
					</ul>
				</div>
				<div className="col-5">
					<h4>Agregar TODO</h4>
					<hr />
						<form onSubmit={handleSubmit}>
							<input 
								className='form-control'
								type="text" 
								name="description"
								autoComplete='false'
								placeholder='Actividad...'
							/>
							<button
								className='btn btn-outline-primary mt-1 btn-block'
								type='submit'
							>
								Agregar
							</button>
						</form>
				</div>
			</div>
		</div>
	)
}
