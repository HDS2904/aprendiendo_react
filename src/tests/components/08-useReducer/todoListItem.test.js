import { fireEvent, render } from "@testing-library/react"
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem"
import { demoTodos } from "../../fixtures/demoTodos"

describe('Pruebas en <TodoListItem />', () => {
	const handleDelete = jest.fn();
	const handleToggle = jest.fn();
	let wrapper;
	
	beforeEach( ()=>{
		wrapper = render( <TodoListItem 
			todo={ demoTodos[1] }
			index={ 0 }
			handleDelete={ handleDelete }
			handleToggle={ handleToggle }
		/> )

	})

	test('Deberia de mostrarse correctamente', () => {

		expect( wrapper.container ).toMatchSnapshot();

	})

	test('Debe de llamar la funci贸n handleDelete', () => {

		const button = wrapper.getByText( "Borrar" );
		fireEvent.click( button );
		expect( handleDelete ).toHaveBeenCalledTimes( 1 ); // Evaluar la cantidad de veces llamada la funci贸n
		expect( handleDelete ).toHaveBeenCalledWith( demoTodos[1].id ); // Evaluar el argumento pasado a la funci贸n

	})

	test('Debe de llamar la funci贸n handleToggle', () => {

		const parr = wrapper.container.querySelector('p');
		fireEvent.click( parr );
		expect( handleToggle ).toHaveBeenCalledTimes( 1 );
		expect( handleToggle ).toHaveBeenCalledWith( demoTodos[1].id );

	})

	test('Debe de mostrar el texto correctamente', () => {

		const parr = wrapper.container.querySelector('p');
		fireEvent.click( parr );
		expect( parr.textContent ).toBe( `1. ${demoTodos[1].desc}` );

	})

	test('Debe de tener la clase complete si el TODO.done = true', () => {

		const todo = demoTodos[1];
		todo.done = true;

		const wrapper = render( <TodoListItem 
			todo={ todo }
		/> )

		const parr = wrapper.container.querySelector('p')
		expect( parr.className ).toBe( 'complete' );

	})

})