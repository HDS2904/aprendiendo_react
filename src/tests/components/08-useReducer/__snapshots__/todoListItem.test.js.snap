// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoListItem /> Deberia de mostrarse correctamente 1`] = `
<div>
  <li
    class="list-group-item"
  >
    <p
      class="false"
    >
      1
      . 
      Aprendiendo Node.js
    </p>
    <button
      class="btn btn-danger"
    >
      Borrar
    </button>
  </li>
</div>
`;
